apps/pkcs12.c

```diff
@@ -642,6 +647,33 @@ int pkcs12_main(int argc, char **argv)
             X509_alias_set1(sk_X509_value(certs, i), catmp, -1);
         }
 
+        /* Load the config file */
+        if ((conf = app_load_config(default_config_file)) == NULL)
+            goto export_end;
+        if (!app_load_modules(conf))
+            goto export_end;
+
+        /* Lets get the config section we are using */
+        if ((cb_attr = NCONF_get_string(conf, "pkcs12", "certBagAttr")) != NULL) {
+            if ((cb_sk = NCONF_get_section(conf, cb_attr)) != NULL) {
+
+                /* Loop over each attribute */
+                for (i = 0; i < sk_CONF_VALUE_num(cb_sk); i++) {
+                    CONF_VALUE *val = sk_CONF_VALUE_value(cb_sk, i);
+                    ASN1_OBJECT *oracletrust = OBJ_txt2obj(val->value, 0);
+                    for (j = 0; j < sk_X509_num(certs); j++) {
+                        // Add a single trust value on every certificate, which will be set on the bag later
+                        X509_add1_trust_object(sk_X509_value(certs, j), oracletrust);
```

Comment:
Another thought is that certs that already have OpenSSL trust markings may already have this trust set on them - regardless of anything that you have in the config file. Which means they would automatically get jdk trust markings too in the pkcs12....not sure we want that?

Reply:
@mattcaswell 
> Another thought is that certs that already have OpenSSL trust markings...

What would those markings be? Does the OpenSSL have its own way of marking certificates in PKCS12 file as trusted?

Reply:
> What would those markings be? Does the OpenSSL have its own way of marking certificates in PKCS12 file as trusted?

It has its own way of marking *certificates* as trusted. It doesn't (currently) transfer those trust markings to/from PKCS12 at the moment. This PR would change that.

For example look at the test cert "ca+serverAuth.pem" in the OpenSSL source distribution:
https://github.com/openssl/openssl/blob/master/test/certs/ca%2BserverAuth.pem

````
$ openssl x509 -in ca+serverAuth.pem -noout -text
...snip...
Trusted Uses:
  TLS Web Server Authentication
No Rejected Uses.
````

Reply:
@mattcaswell 
Maybe a dumb question but its Friday evening and I am tired, are you referring to the text:
```
-----BEGIN TRUSTED CERTIFICATE-----
```
As a way of saying it is trusted, or something in its ASN.1 encoding (and if latter, what)?

Reply:
That header indicates it has some form of OpenSSL trust settings encoded into the file. The command line output I showed above in my comment tells you what those trust markings are, i.e. this cert is trusted for "TLS Web Server Authentication".

Reply:
Those do not seem to be encoded to any standard:

![image](https://user-images.githubusercontent.com/16415478/185667788-1ab0123f-224c-47c6-a95b-805495eb10e7.png)

If the certificate was trusted for `TLS Web Server Authentication` shouldn't that show under `extendedKeyUsage`?

I mean, I do see it in ASN.1 dump:

![image](https://user-images.githubusercontent.com/16415478/185668005-700831d0-2c4f-421d-b6f8-91faf3e66e4c.png)

But it does not seem encoded as part of standard X.509?

Reply:
Why? If they do have other OpenSSL trust markings, without the JDK trust, then I'm guessing java wouldn't be happy with the pkcs12? 

Do you know how I could create a public cert with other trust markings and see what happens when I add them to a pkcs12 (without my change) and see what Keytool thinks of them and whether java can import them into a TrustStore?

Reply:
`For example look at the test cert "ca+serverAuth.pem" in the OpenSSL source distribution:
https://github.com/openssl/openssl/blob/master/test/certs/ca%2BserverAuth.pem`

hmm that's interesting, if I use that certificate as my input certificate when exporting the PKCS12 then my openssl pkcs12 -export... fails with

"Error creating PKCS12 structure.."...wonder what my code is doing to break this.

Reply:
https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.12 mentions nothing about the "Trusted Uses"

Reply:
https://www.openssl.org/docs/man1.0.2/man1/x509.html ah this mentions them

Reply:
Ah I see, you create a normal cert+key pair, then run openssl x509 and use -addtrust clientAuth...that adds the TrustedUses:
  TLS Web Client Authentication

Reply:
> But it does not seem encoded as part of standard X.509?

It's not. Well, not exactly, anyway. An OpenSSL "TRUSTED CERTIFICATE" contains a normal X.509 certificate concatenated with an OpenSSL X509_CERT_AUX structure on the end:

https://github.com/openssl/openssl/blob/6a9255054b345026bc847ddad72f2da93f30ce4c/crypto/x509/x_x509a.c#L26-L32

> If the certificate was trusted for TLS Web Server Authentication shouldn't that show under extendedKeyUsage?

While the OpenSSL trust markings use the same extendedKeyUsage values as you might see encoded *inside* an X509 structure, the OpenSSL trust markings are *external* to the certificate itself.

The difference is between who is asserting what the certificate can be used for. The CA that issues the cert asserts what the certificate is to be used for via the extendedKeyUsage in the cert itself. However the maintainer of the trust store may override those with their own trust settings.

For example the CA could state that a certificate can be used for TLS Web Server Authentication and TLS Web Client Authentication. The maintainer of the trust store can put the cert in the store but through trust markings indicate that they actually only want to trust it for TLS Web Server Authentication and not TLS Web Client Authentication.

Reply:
> I guess what I am trying to point out is that other programs (well at least Windows cert manager) do not recognize that way of trusting a certificate for certain usage.

Right the OpenSSL trust markings are an OpenSSL mechanism only. They don't get exported to pkcs12 (at least they didn't before this PR).

Actually the "jdkTrustedKeyUsage" mechanism is kind of analogous to the OpenSSL mechanism. Its seems to be a mechanism for marking trust externally to the cert itself, i.e. its held as an attribute of the PKCS12_SAFEBAG which separately holds the cert.

This PR is (ab)using the OpenSSL trust marking system to store key usage details alongside an X509 cert and then mapping that into the jdkTrustedKeyUsage attribute in the PKCS12_SAFEBAG. It's actually not an unreasonable thing to do since the two mechanisms seem very similar. The problem is I see potential issues where OpenSSL certs already have these trust markings and they are not intended to be exported to pkcs12 - but this PR would mean that they are exported anyway.

Reply:
@mattcaswell 

I see.

I think there is one crucial difference between OpenSSL trust and jdkTrustedKeyUsage mechanism which has more to do with the file format itself but it still seems important.

Correct me if I am wrong, but OpenSSL trust does not seem to be authenticated in any way? You can just append it to the certificate to mark it as trusted and anyone with file access can add, modify, or remove this trust, right?

On the other hand PKCS12 safebag attribute cannot be added without authenticating to the PKCS12 certificate store using a password (which is optional, but at least it exists as an option).

I don't know about others who participated in the feature request that led to this PR, but I personally wouldn't want OpenSSL trust to be mapped to `jdkTrustedKeyUsage`. JDK usage so far can only be full trust or no trust, so attempting to map that to OpenSSL granular control doesn't really make sense to me.

If anything, I would prefer OpenSSL trust stripped from the certificate when it gets stored in a PKCS12 file since it is not standard X.509, and PKCS12 is also used as a certificate + key distribution and exchange format not just as a trust store so stricter adherence to standards should be the goal.

At the minimum, removing OpenSSL trust on import should exist as an option if it doesn't already.

Reply:
> Correct me if I am wrong, but OpenSSL trust does not seem to be authenticated in any way? You can just append it to the certificate to mark it as trusted and anyone with file access can add, modify, or remove this trust, right?

Correct. It is assumed that the trust store is protected by file system controls and that anyone with the access rights to write to the file is sufficiently authorised to manage those trusts.

> If anything, I would prefer OpenSSL trust stripped from the certificate when it gets stored in a PKCS12 file

Currently OpenSSL trust does not get imported into a PKCS12 or exported from it. But because this PR is (ab)using the OpenSSL trust mechanism it would actually change that so that OpenSSL's trust would get mapped to `jdkTrustedKeyUsage`. So I'm not convinced that the strategy in this PR is quite right.

Reply:
@mattcaswell 
> So I'm not convinced that the strategy in this PR is quite right.

To be honest me neither, especially because nobody in the feature request asked for that mapping to happen.

@grahamwoodward I agree with @mattcaswell here, I didn't look at your code, but this mapping should not be happening. We don't want to inherit OpenSSL trust and turn it into jdkTrustedKeyUsage.

Reply:
@levicki @mattcaswell yeah that's fine, I'm happy to change it I'm just trying to process/understand what you're saying with how it should be

Reply:
am I right in thinking this part of the PR isn't needed at all?

Ref the other comments below, do I just need the callback to read the parse the cnf, find the new section and lookup the values in there and add to the bag attr?

Reply:
Right - if you go the callback route then you wouldn't need to set trust at all on the X509 objects and so this part of the PR could be removed.

> Ref the other comments below, do I just need the callback to read the parse the cnf, find the new section and lookup the values in there and add to the bag attr?

Probably you would read/parse the cnf outside of the callback and store details in some global variable for the app. The callback would be called multiple times - once for each cert, and would just add the appropriate attributes as needed.